source/credentials_provider_sts.c

```diff
@@ -0,0 +1,668 @@
+
+    AWS_LOGF_DEBUG(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "static: creating STS credentials provider");
+    if (!provider) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*provider);
+    AWS_ZERO_STRUCT(*impl);
+
+    aws_credentials_provider_init_base(provider, allocator, &s_aws_credentials_provider_sts_vtable, impl);
+
+    impl->function_table = &s_default_function_table;
+
+    if (options->function_table) {
+        impl->function_table = options->function_table;
+    }
+
+    if (options->tls_ctx) {
+        AWS_LOGF_TRACE(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): tls context provided, using pre-built tls context.",
+            (void *)provider);
+        impl->ctx = options->tls_ctx;
+    } else {
+        AWS_LOGF_TRACE(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): tls context not provided, initializing a new one",
+            (void *)provider);
+        struct aws_tls_ctx_options tls_options;
+        aws_tls_ctx_options_init_default_client(&tls_options, allocator);
+        impl->ctx = aws_tls_client_ctx_new(allocator, &tls_options);
+
+        if (!impl->ctx) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p): failed to create a tls context with error %s",
+                (void *)provider,
+                aws_error_debug_str(aws_last_error()));
+            aws_tls_ctx_options_clean_up(&tls_options);
+            goto cleanup_provider;
+        }
+
+        impl->owns_ctx = true;
+    }
+
+    if (!options->creds_provider) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p): A credentials provider must be specified", (void *)provider);
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        goto cleanup_provider;
+    }
+
+    impl->role_session_name =
+        aws_string_new_from_array(allocator, options->session_name.ptr, options->session_name.len);
+
+    if (!impl->role_session_name) {
+        goto cleanup_provider;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): using session_name %s",
+        (void *)provider,
+        aws_string_c_str(impl->role_session_name));
+
+    impl->assume_role_profile = aws_string_new_from_array(allocator, options->role_arn.ptr, options->role_arn.len);
+
+    if (!impl->assume_role_profile) {
+        goto cleanup_provider;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): using assume_role_arn %s",
+        (void *)provider,
+        aws_string_c_str(impl->assume_role_profile));
+
+    impl->duration_seconds = options->duration_seconds;
+
+    if (impl->duration_seconds < 900) {
+        impl->duration_seconds = 900;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): using credentials duration %" PRIu16,
+        (void *)provider,
+        impl->duration_seconds);
+
+    impl->provider = options->creds_provider;
```

Comment:
need an acquire here.  Given that there's a release in the clean up and this doesn't crash I'm guessing that the caller of this is missing a release.